src/Common/ThreadPool.cpp

```diff
@@ -180,6 +183,11 @@ ReturnType ThreadPoolImpl<Thread>::scheduleImpl(Job job, Priority priority, std:
             return false;
     };
 
+    double schedule_fault_probability = getFunctionFaultProbability(FaultFunctionType::ThreadPoolSchedule);
+    std::bernoulli_distribution fault(schedule_fault_probability);
+    if (unlikely(schedule_fault_probability >= 0) && fault(thread_local_rng))
+        return static_cast<ReturnType>(on_error("no free thread fault"));
+
```

Comment:
Have you thought about hiding the details inside `FunctionsFuzzer`? Maybe with a generic function something like:
```
if (FunctionsFuzzer::shouldFail(FaultFunctionType::ThreadPoolSchedule))
    return static_cast<ReturnType>(on_error("no free thread fault"));
```

This way in `FucntionsFuzzer::shouldFail` we first check the probability and if the probability is `0`, then we can early return false avoiding initializing the and generating random bool generator and generating "always" false.

I think this would also make it eaiser to reuse `FunctionsFuzzer` without duplicating this logic.